### DISCUSSION TITLE
Can handle NetBeans all textmate features?

### DISCUSSION THREAD

**Discussion by Chris2011** [apache/netbeans#7852] [Q&A]
It came into my mind, when I wanted to add vue into NetBeans completely over textmate for syntax highlighting and LSP for vue features, etc. The problem here is, it seems that NetBeans don't handle includes or sub languages from textmate.

Vue can handle html (default), pug, haml, ts, js, css, scss, less, stylus, postcss. And if some pattern is matching for example lang="pug" there is this block in the textmate file:

```json
"patterns": [
  {
      "include": "#tag-stuff"
  },
  {
      "contentName": "text.pug",
      "begin": "(>)",
      "beginCaptures": {
          "1": {
              "name": "punctuation.definition.tag.end.html"
          }
      },
      "end": "(?=</template>)",
      "patterns": [
          {
              "include": "text.pug"
          }
      ]
  }
]
```

and when I understand it correctly, it searches for a registered pug textmate module, but how is this done in NetBeans? Is this already implemented and if yes, how this is working? Maybe @jlahoda or @matthiasblaesing can help here. Thx.

**Comment by matthiasblaesing** [comment]
My understanding of this:

https://github.com/apache/netbeans/blob/70f69415a279cba3a6ae06345a114a8ec518c4e2/ide/textmate.lexer/src/org/netbeans/modules/textmate/lexer/TextmateTokenId.java#L119-L154

runs through all "config" file/objects under the `Editors` branch in the system file system. It will pick up all files, that have an `textmate-grammar` or `inject-to` attribute. For your usecase you are looking for the former. For C this looks like this:

![grafik](https://github.com/user-attachments/assets/39d44653-ff50-438d-b912-21f84abf5cd0)

In the same file you find:

https://github.com/apache/netbeans/blob/70f69415a279cba3a6ae06345a114a8ec518c4e2/ide/textmate.lexer/src/org/netbeans/modules/textmate/lexer/TextmateTokenId.java#L161-L183

This is the code that makes the textmake grammar/lexer definitions available to NetBeans and as you can see it uses the `scope2File` map created by the code mentioned above to find the grammar files for the provided `scope` name.

The entry from the graphic above is created by this:

https://github.com/apache/netbeans/blob/70f69415a279cba3a6ae06345a114a8ec518c4e2/cpplite/cpplite.editor/src/org/netbeans/modules/cpplite/editor/file/CDataObject.java#L117

The annotation processor in the `textmate.lexer` module compiles that into a `layer.xml` and that in turn results in the tree entry you see.

So register the grammar and you should be done.

**Reply by Chris2011** [reply]
Thanks for your info. So I need to register all supported embedded languages for the vue mimetype that I will create? So HTML, PUG, TS, SASS, LESS, etc.? They need to be registered before, right?

**Reply by matthiasblaesing** [reply]
Yes that would be my understanding.

**Reply by Chris2011** [reply]
Oki, will check this. Thx.
